waddrmgr/scoped_manager.go

```diff
@@ -579,6 +593,75 @@ func (s *ScopedKeyManager) AccountProperties(ns walletdb.ReadBucket,
 	return props, nil
 }
 
+// cachedKey is an entry within the LRU map that stores private keys that are
+// to be used frequently. We use this wrapper struct to be able too report the
+// size of a given element to the cache.
+type cachedKey struct {
+	key *btcec.PrivateKey
+}
+
+// Size returns the size of this element. Rather than have the cache limit
+// based on bytes, we simply report that each element is of size 1, meaning we
+// can set our cached based on the amount of keys we want to store, rather than
+// the total size of all the keys.
+func (c *cachedKey) Size() (uint64, error) {
+	return 1, nil
+}
+
+// DeriveFromKeyPathCache is identical to DeriveFromKeyPath, however it'll fail
+// if the account refracted in the DerivationPath isn't already in the
+// in-memory cache. Callers looking for faster private key retrieval can opt to
+// call this method, which may fail if things aren't in the cache, then fall
+// back to the normal variant.
+func (s *ScopedKeyManager) DeriveFromKeyPathCache(kp DerivationPath,
+) (*btcec.PrivateKey, error) {
+
+	s.mtx.Lock()
+	defer s.mtx.Unlock()
+
+	// First, try to look up the key itself in the proper cache, if the key
+	// is here, then we don't need to do anything further.
+	privKeyVal, err := s.privKeyCache.Get(kp)
+	if err == nil {
+		return privKeyVal.(*cachedKey).key, nil
+	}
+
+	// If the key isn't already in the cache, then we'll try to look up the
+	// account info in the cache, if this fails, then we exit here as we
+	// can't move forward without creating a DB transaction, and the point
+	// of this method is to avoid that.
+	acctInfo, ok := s.acctInfo[kp.InternalAccount]
```

Comment:
Ah, I wasn't aware that we already have this account cache that also has the extended account key. I assume getting rid of two levels of derivation is performance gain enough to warrant the additional private key cache? Or is getting rid of the DB transaction already enough and we could just do the derivation each time instead? Just thinking aloud here...

Reply:
> Ah, I wasn't aware that we already have this account cache that also has the extended account key. I assume getting rid of two levels of derivation is performance gain enough to warrant the additional private key cache? 

The idea here is just skip all the extra derivation all together since it's a pretty easy win. If we cached the account info only, we'd still need to re-derive actual final key which means the extra EC+hashing operations. 

> I don't really see a reason to separate the two caches except for maybe code simplicity, since on retrieval you'd have to go by account first then attempt to get by index

One cache is just maintaining the account information, and the top-level account key for a given key family. This cache saves us from the expensive `scrypt` operations to re-derive the encryption key to decrypt the master private key, but doesn't let us skip the extra BIP-32 operations needed to derive the actual key we need. 

Take a look at `loadAccountInfo` to see al the operations we skip by checking this cache, and `deriveKey` for the operations we skip by caching the key itself. Why have the overhead of a series of asymmetric crypto operations each time we need to: sign a new state, create an invoice, decode error responses?